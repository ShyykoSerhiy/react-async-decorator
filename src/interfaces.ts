import { TSyncPromise } from './promise';
export interface Action<T = any> {
	type: T;
}

export interface AnyAction extends Action {
	// Allows any extra properties to be defined in an action.
	[extraProps: string]: any;
}

export interface Dispatch<A extends Action = AnyAction> {
	<T extends A>(action: T): T;
}

export interface IActionFetchClear extends Action {
	name: string;
	action: 'clear';
}

export interface IActionFetchError extends Action {
	name: string;
	action: 'error';
	key: string;
	error: Error;
}

export interface IActionFetchSet extends Action {
	name: string;
	action: 'set';
	key: string;
	value: any;
}

export interface IActionFetchRequest extends Action {
	name: string;
	action: 'request';
	key: string;
}

export type IActionFetch = IActionFetchClear | IActionFetchSet | IActionFetchRequest | IActionFetchError;

export interface IDataItem {
	data?: any;
	error?: {
		name: string;
		message: string;
		stack: string;
	};
	loading?: boolean;
}

export type FetcherItem = Partial<Record<string, IDataItem>>;
export type FetcherState = Partial<Record<string, FetcherItem>>;

export interface MiddlewareAPI<D extends Dispatch = Dispatch<any>, S = Record<string, FetcherState>> {
	dispatch: D;
	getState(): S;
}

export interface IOption {
	name: string;
	action: string; // action for redux
	key: string; // key in store
  store: MiddlewareAPI;
  hashArg?<T>(arg?: T): string;
}


export type AnyResult<T> = T | Promise<T> | TSyncPromise<T> | Promise<T>[] | TSyncPromise<T>[];

export interface IHolder<T> {
  set(key: string, defer: TSyncPromise<T>): void;
  allow(key: string, defer: TSyncPromise<T>): boolean;
  clear(): void;
  get(key: string): T | undefined;
  has(key: string): boolean;
  isLoading(key: string): boolean | undefined;
  error(key: string): Error | undefined;
  getAwait(key: string): TSyncPromise<T> | undefined;
  await(key: string): TSyncPromise<T>;
  awaitAll(): TSyncPromise<T[]>;
}

export interface IFetcherContext<T> {
  holder: IHolder<T>;
}
export interface IFetcherFunction<T, GetOptions, SetOptions> {
  load?: IFetcherFnContext<GetOptions, AnyResult<T>>;
  modify?: IFetcherFnContext<SetOptions, AnyResult<T>>;
}

export interface IFetcherBase<SetOptions, T> {
	clear(): void;
	await(): TSyncPromise<T>;
	awaitAll(): TSyncPromise<T[]>;
	asyncSet(opts: SetOptions): TSyncPromise<T>;
	implModify(modify: IFetcherFnContext<SetOptions, AnyResult<T> | undefined>): void;
	isLoading(): boolean;
}

export type IFetcherFn<Opt, T> = (arg?: Opt) => T;
export type IFetcherFnContext<Opt, T> = (ctx: IFetcherContext<T>, arg?: Opt) => T;

export interface IFetcher<T, GetOptions = any, SetOptions = any> extends IFetcherBase<SetOptions, T> {
	impl(load: IFetcherFnContext<GetOptions, AnyResult<T>>): void;
	asyncGet: IFetcherFn<GetOptions, TSyncPromise<T>>;
	get: IFetcherFn<GetOptions, T>;
}

export interface IFetcherOption {
  name?: string;
  setItem?(action: IActionFetch): Record<string, any> | undefined;
}
