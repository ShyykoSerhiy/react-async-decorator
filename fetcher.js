var Holder = require('./holder').Holder;

function Fetcher(opts) {
    this.load = opts.load;
    this.holder = new Holder(opts);
}

Fetcher.prototype.clear = function () {
    return this.holder.clear();
};

Fetcher.prototype.awaitAll = function() {
    return this.holder.awaitAll();
}

Fetcher.prototype.await = function() {
    var args = Array.prototype.slice.apply(arguments);
    var key = args.length ? JSON.stringify(args) : '';
    return this.holder.await(key);
}

Fetcher.prototype.asyncGet = function() {
    var args = Array.prototype.slice.apply(arguments);
    var key = args.length ? JSON.stringify(args) : '';
    this.init(key);
    return this.holder.await(key);
}

Fetcher.prototype.init = function(key) {
    if (!this.holder.has(key)) {
        var defer = this.load.apply(null, arguments);
        this.holder.set(key, defer);
    }
}

Fetcher.prototype.get = function () {
    var args = Array.prototype.slice.apply(arguments);
    var key = args.length ? JSON.stringify(args) : '';
    this.init(key);
    var error = this.holder.error(key);
    if (error !== undefined) {
        throw error;
    }
    if (!this.holder.has(key)) {
        throw this.holder.await(key);
    }
    return this.holder.get(key);
};

var counter = 0;

function create(accessor) {
    function createFetcher(load, name) {
        var pname = (name === undefined || name === null) ? ('' + counter++) : name;
        return new Fetcher({ load: load, name: pname, accessor: accessor });
    }
    return createFetcher;
}

module.exports.Fetcher = Fetcher;
module.exports.create = create;
