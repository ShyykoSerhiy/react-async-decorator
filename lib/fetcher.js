var Holder = require('./holder').Holder;

function notImpl() {
    var msg = new Error('Fetcher wasn\'t implemented');
    return Promise.reject(msg);
}

function Fetcher(opts) {
    this.load = opts.load || notImpl;
    this.holder = new Holder(opts);
}

Fetcher.prototype.impl = function(load) {
    this.load = load || notImpl;
    this.holder.clear();
}

Fetcher.prototype.clear = function () {
    return this.holder.clear();
};

Fetcher.prototype.awaitAll = function() {
    return this.holder.awaitAll();
}

Fetcher.prototype.await = function() {
    var args = Array.prototype.slice.apply(arguments);
    var key = args.length ? JSON.stringify(args) : '';
    return this.holder.await(key);
}

Fetcher.prototype.asyncGet = function() {
    var args = Array.prototype.slice.apply(arguments);
    var key = args.length ? JSON.stringify(args) : '';
    this.init(key, args);
    return this.holder.await(key);
}

Fetcher.prototype.init = function(key, args) {
    if (!this.holder.has(key)) {
        var defer = this.load.apply(null, args);
        this.holder.set(key, defer);
    }
}

Fetcher.prototype.get = function () {
    var args = Array.prototype.slice.apply(arguments);
    var key = args.length ? JSON.stringify(args) : '';
    this.init(key, args);
    var error = this.holder.error(key);
    if (error !== undefined) {
        throw error;
    }
    if (!this.holder.has(key)) {
        throw this.holder.await(key);
    }
    return this.holder.get(key);
};

function create(accessor) {
    var counter = 0;
    function createFetcher(load, name) {
        var pname = (name === undefined || name === null) ? ('' + counter++) : name;
        return new Fetcher({ load: load, name: pname, accessor: accessor });
    }
    return createFetcher;
}

module.exports.Fetcher = Fetcher;
module.exports.create = create;
