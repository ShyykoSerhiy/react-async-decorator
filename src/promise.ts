export type TCallback<T> = (
  resolve: (data: T) => void,
  reject: (err: Error) => void
) => void;

interface IData<Type, T> {
  type: Type;
  data: T;
}

export class Iterator<T extends ArrayLike<any>> {
  private data: T = [] as any;
  private error: Error | undefined;
  private executed = false;

  constructor(
      private count: number,
      private onSuccess: (data: T) => void,
      private onFail: (data: Error) => void
  ) {}

  public handleSuccess = (data: any, index: number): void => {
      (this.data as any)[index] = data;
      this.count--;
      if (this.count <= 0 && !this.executed) {
          this.executed = true;
          this.onSuccess(this.data);
      }
  }
  public handleError = (error: Error): void => {
      if (!this.executed && !this.error ) {
          this.error = error;
          this.executed = true;
          this.onFail(this.error);
      }
  }
}


export class TSyncPromise<T> implements Promise<T> {
  public data?: IData<'resolved', T> | IData<'rejected', any>;
  private onfulfilled?: Array<(value: T) => any>;
  private onrejected?: Array<(err: any) => any>;

  [Symbol.toStringTag]: 'Promise';

  static resolve<O>(data: O | PromiseLike<O>): TSyncPromise<O> {
    if (data instanceof Promise) {
      return new TSyncPromise<O>((resolve, reject) => data.then(resolve, reject))
    } else if (data && data.toString && data.toString() === 'TSyncPromise') {
      return data as any;
    } else {
      const p = new TSyncPromise<O>(null as any);
      p.data = { type: 'resolved', data: data as any };
      return p;
    }
  }

  static reject<O>(data: any): TSyncPromise<O> {
    const p = new TSyncPromise<O>(null as any);
    p.data = { type: 'rejected', data: data as any };
    return p;
  }

  static all<A1, A2, A3, A4, A5, A6, A7>(
    p: [Promise<A1> | A1, Promise<A2> | A2, Promise<A3> | A3, Promise<A4> | A4, Promise<A5> | A5, Promise<A6> | A6, Promise<A7> | A7]
  ): TSyncPromise<[A1, A2, A3, A4, A5, A6, A7]>;
  static all<A1, A2, A3, A4, A5, A6>(
    p: [Promise<A1> | A1, Promise<A2> | A2, Promise<A3> | A3, Promise<A4> | A4, Promise<A5> | A5, Promise<A6> | A6]
  ): TSyncPromise<[A1, A2, A3, A4, A5, A6]>;
  static all<A1, A2, A3, A4, A5>(
    p: [Promise<A1> | A1, Promise<A2> | A2, Promise<A3> | A3, Promise<A4> | A4, Promise<A5> | A5]
  ): TSyncPromise<[A1, A2, A3, A4, A5]>;
  static all<A1, A2, A3, A4>(
    p: [Promise<A1> | A1, Promise<A2> | A2, Promise<A3> | A3, Promise<A4> | A4]
  ): TSyncPromise<[A1, A2, A3, A4]>;
  static all<A1, A2, A3>(
    p: [Promise<A1> | A1, Promise<A2> | A2, Promise<A3> | A3]
  ): TSyncPromise<[A1, A2, A3]>;
  static all<A1, A2>(
    p: [Promise<A1> | A1, Promise<A2> | A2]
  ): TSyncPromise<[A1, A2]>;
  static all<T>(p: TSyncPromise<T>[] | Promise<T>[] | T[]): TSyncPromise<T[]>;
  static all(p: any[]): TSyncPromise<any[]> {
    return new TSyncPromise<any[]>((resolve, reject) => {
      if (!p.length) {
        return resolve([]);
      }
      const iterator = new Iterator(p.length, resolve, reject);
      p.forEach((arg, index) => {
        if (arg instanceof Promise || (arg.toString && arg.toString() === 'TSyncPromise')) {
          arg.then(d => iterator.handleSuccess(d,index), iterator.handleError);
        } else {
          iterator.handleSuccess(arg, index);
        }
      });
    });
  }

  public constructor(fn: TCallback<T>) {
    if (fn) {
      fn(
        data => this.setData('resolved', data),
        data => this.setData('rejected', data)
      );
    }
  }

  private setData(type: 'resolved' | 'rejected', data: any) {
    if (!this.data) {
      this.data = { type, data } as any;
      if (type === 'resolved' && this.onfulfilled) {
        this.onfulfilled.forEach(cb => cb(data));
      } else if (type === 'rejected' && this.onrejected) {
        this.onrejected.forEach(cb => cb(data));
      }
      this.onfulfilled = undefined;
      this.onrejected = undefined;
    }
  }

  public then<TResult1 = T, TResult2 = never>(
    onfulfilled?: null | ((value: T) => TResult1 | PromiseLike<TResult1>),
    onrejected?: null | ((reason: any) => TResult2 | PromiseLike<TResult2>)
  ): TSyncPromise<TResult1 | TResult2> {
    const d = this.data;
    if (d) {
      if (onfulfilled && d.type === 'resolved') {
        const p = onfulfilled(d.data)
        return TSyncPromise.resolve(p);
      } else if (onrejected && d.type === 'rejected') {
        const p = onrejected(d.data);
        return TSyncPromise.resolve(p);
      } else {
        return this as any;
      }
    } else {
      const fulfilled = this.onfulfilled || (this.onfulfilled = []);
      const rejected = this.onrejected || (this.onrejected = []);
      return new TSyncPromise<TResult1 | TResult2>((resolve, reject) => {
        fulfilled.push(resolve as any);
        rejected.push(reject);
        if (onfulfilled) {
          fulfilled.push(onfulfilled)
        }
        if (onrejected) {
          rejected.push(onrejected);
        }
      });
    }
  }
  public catch<TResult = never>(
    onrejected?: (reason: any) => TResult | PromiseLike<TResult>
  ): Promise<T | TResult> {
    return this.then(undefined, onrejected);
  }

  public toPromise() {
    return new Promise<T>((resolve, reject) => this.then(resolve, reject));
  }
  toString() {
    return 'TSyncPromise';
  }
}
