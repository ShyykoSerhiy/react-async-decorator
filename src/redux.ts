import { create } from './fetcher';
import { MiddlewareAPI, Dispatch, IDataItem, IActionFetch, typeFetcherFn } from './interfaces';
import { IOptionStore, createReducer, IOptionReducer } from './reduxReducer';
export { MiddlewareAPI, Dispatch, IDataItem, IActionFetch };

export function initRedux<State>() {
	function createReduxFetcher(opt: IOptionStore) {
		let api: MiddlewareAPI;

		function use(aApi: MiddlewareAPI): void {
			api = aApi;
		}

		const store: MiddlewareAPI = {
			getState() {
				return api.getState();
			},
			dispatch(action: any) {
				return api.dispatch(action);
			},
		};

    // reducer initialize after create call
    let reducer: (state: State, act: IActionFetch) => State;

    // it's sync callback which call after create call
    function createStore(opt: IOptionReducer<any>): MiddlewareAPI {
      reducer = createReducer<State>(opt);
      return store;
    }

		const fn = create({ ...opt, createStore }) as any;
		return { use, reducer, createFetcher: fn as typeof typeFetcherFn };
	}
	return createReduxFetcher;
}
