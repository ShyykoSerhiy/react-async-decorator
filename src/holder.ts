import {
	IOption,
	IDataItem,
	IHolder,
  IActionsLifecycle,
  IOptionActions,
  IActionFetchSet,
  IActionFetchError,
  IActionFetch
} from './interfaces';
import { TSyncPromise } from './promise';

export function createActions<T>(opts: IOptionActions, holder: Holder<T>): IActionsLifecycle<T> {
  const { store, name, action } = opts;
  const actions: IActionsLifecycle<T> = {
    success(keys: string[], params: Record<string, TSyncPromise<T>>, payloads: T[]) {
      const allowKeys = keys.filter(key => holder.allow(key, params[key]));
      const indexes = allowKeys.map(key => keys.indexOf(key));
      const payload: Record<string, T> = {};
      indexes.forEach(id => {
        payload[keys[id]] = payloads[id];
      });
      if (indexes.length) {
        const act2: IActionFetchSet = {
          type: action,
          action: 'set',
          name,
          payload,
        };
        store.dispatch(act2);
      }
    },
    error(keys: string[], params: Record<string, TSyncPromise<T>>, error: Error) {
      const allowKeys = keys.filter(key => holder.allow(key, params[key]));
      const indexes = allowKeys.map(key => keys.indexOf(key));
      const payload: Record<string, Error> = {};
      indexes.forEach(id => {
        payload[keys[id]] = error;
      });
      const act2: IActionFetchError = {
        type: action,
        action: 'error',
        name,
        payload,
      };
      store.dispatch(act2);
    },
    request(keys: string[]) {
      const act: IActionFetch = { type: action, action: 'request', name, keys };
      store.dispatch(act);
    }
  };
  return actions;
}

export class Holder<T> implements IHolder<T> {
  private container: Partial<Record<string, TSyncPromise<T>>> = {};
	static notExist = TSyncPromise.reject<any>(new Error("Doesn't exist"));
  public actions: IActionsLifecycle<T>;
	constructor(private props: IOption) {
    this.actions = createActions(props, this);
  }

	set(params: Record<string, TSyncPromise<T>>): void;
	set(key: string, defer: TSyncPromise<T>): void;
	set(aKey: any, aDefer?: any) {

    const isPropsArg = aDefer !== undefined;
    const keys = isPropsArg ? [aKey] : Object.keys(aKey);
    const params: Record<string, TSyncPromise<T>> = isPropsArg ? { [aKey]: aDefer } : aKey;
    const defers = keys.map(key => {
      this.container[key] = params[key];
      return params[key];
    });
    if (!this.props.manual) {
      this.actions.request(keys);
      TSyncPromise.all<T>(defers).then(
        payloads => this.actions.success(keys, params, payloads),
        error => this.actions.error(keys, params, error)
      );
    }
	}

	allow(key: string, defer: TSyncPromise<T>): boolean {
		return this.container[key] === defer;
	}

	clear(): void {
		const { store, action, name } = this.props;
		store.dispatch({ type: action, action: 'clear', name });
		this.container = {};
	}

	private getDataBlob(aKey: string): IDataItem | undefined {
		const { store, key, name } = this.props;

		const state = store.getState();
		if (!state) {
			return;
		}
		const holderState = state[key];
		if (!holderState) {
			return;
		}
		const ptr = holderState[name];
		if (!ptr) {
			return;
		}
		return ptr[aKey];
	}

	get(key: string): T | undefined {
		const blob = this.getDataBlob(key);
		return blob ? blob.data : undefined;
	}

	has(key: string): boolean {
		const blob = this.getDataBlob(key);
		return blob ? (blob.loading === true ? false : true) : false;
	}

	isLoading(key: string): boolean | undefined {
		const blob = this.getDataBlob(key);
		return blob ? blob.loading : undefined;
	}

	error(key: string): Error | undefined {
		const blob = this.getDataBlob(key);
		const error = blob ? blob.error : undefined;
		if (error) {
			const err = new Error();
			err.message = error.message;
			err.stack = error.stack;
			err.name = error.name;
			return err;
		}
	}

	getAwait(key: string): TSyncPromise<T> | undefined {
		return this.container[key];
	}

	await(key: string): TSyncPromise<T> {
		const ret = this.getAwait(key);
		return ret || Holder.notExist;
	}

	awaitAll(): TSyncPromise<T[]> {
		const allDefers = Object.keys(this.container).map(key => this.container[key]);
		return TSyncPromise.all<T>(allDefers);
	}
}
