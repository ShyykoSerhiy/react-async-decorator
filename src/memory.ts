import { MiddlewareAPI, FetcherState, FetcherItem, TContainer } from './interfaces';
import { IOptionReducer, createReducer } from './reduxReducer';
import { TSyncPromise } from './promise';

export function createMemoryStore(opt: IOptionReducer<any>, initState: FetcherState = {}): MiddlewareAPI {
	const { key } = opt;
	const reducer = createReducer(opt);
	let state: any = { [key]: initState };
	const ret: MiddlewareAPI = {
		getState() {
			return state;
		},
		dispatch(action: any): any {
			state = reducer(state, action);
			return action;
		},
	};
	return ret;
}

export function snapshotStoreFetcher(holderState: FetcherItem): TContainer<any> {
	return Object.keys(holderState).reduce((m2, name) => {
		const ptr = holderState[name];
		if (ptr) {
			m2[name] = new TSyncPromise(resolve => resolve(ptr.data));
		}
		return m2;
	}, {});
}

export function snapshotStore(
	state: FetcherState,
	container: Record<string, TContainer<any>> = {}
): Partial<Record<string, TContainer<any>>> {
	return Object.keys(state).reduce((memo, key) => {
		const holderState = state[key];
		if (!holderState) {
			memo[key] = snapshotStoreFetcher(holderState);
		}
		return memo;
	}, container);
}
