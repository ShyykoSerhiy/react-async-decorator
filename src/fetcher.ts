import { IOption, MiddlewareAPI, FetcherState, AnyResult, IFetcherFunction, typeFetcherFn, IActionFetch } from './interfaces';
import { Holder } from './holder';
import { TSyncPromise } from './promise';
import { IOptionReducer, createReducer } from './reduxReducer';
function notImpl() {
	const msg = new Error("Fetcher wasn't implemented");
	return TSyncPromise.reject(msg);
}

export class Fetcher<SetOptions, T> {
	private load: (...args: any[]) => AnyResult<T>;
	private modify: (opt: SetOptions) => AnyResult<T>;
	private holder: Holder<T>;
	private mapper?: Partial<Record<string, Fetcher<SetOptions, T>>>;
	private keyPersist: string = '';

	constructor(opts: IOption, fn: IFetcherFunction<T>) {
		this.holder = new Holder<T>(opts);
		const impl = fn.load || notImpl;
		this.impl(impl as any);
	}

	impl(load: (...args: any[]) => AnyResult<T>): void {
		this.load = load;
	}

	implModify(modify: (opt: SetOptions) => AnyResult<T> | undefined): void {
		this.modify = modify;
	}

	clear(): void {
		return this.holder.clear();
	}

	init(key: string, args: any[]): void {
		const load = this.load;
		if (!load) {
			return;
		}
		if (this.holder.has(key)) {
			return;
		}
		if (!this.holder.getAwait(key)) {
			const defer = load(...args);
			const syncDefer = (Array.isArray(defer) ? TSyncPromise.all(defer) : TSyncPromise.resolve(defer)) as TSyncPromise<
				T
			>;

			this.holder.set(key, syncDefer);
		}
	}

	isLoading(): boolean {
		return this.holder.isLoading(this.keyPersist);
	}

	awaitAll(): TSyncPromise<T[]> {
		return this.holder.awaitAll();
	}

	await(): TSyncPromise<T> {
		const args = Array.prototype.slice.apply(arguments);
		return this.holder.await(this.keyPersist);
	}

	asyncGet(...args: any[]): TSyncPromise<T> {
		this.init(this.keyPersist, args);
		return this.holder.await(this.keyPersist);
	}

	store(key: string): Fetcher<SetOptions, T> {
		const keyPersist = this.keyPersist || '';
		const mapper = this.mapper || (this.mapper = { [keyPersist]: this });
		if (mapper[key]) {
			return mapper[key];
		} else {
			const copy: Fetcher<SetOptions, T> = Object.create(this);
			copy.keyPersist = key;
			mapper[key] = copy;
			return copy;
		}
	}

	get(...args: any[]): T {
		const key = this.keyPersist;
		this.init(key, args);
		const error = this.holder.error(key);
		if (error !== undefined) {
			throw error;
		}
		if (!this.holder.has(key)) {
			throw this.holder.await(key);
		}
		return this.holder.get(key);
	}

	asyncSet(opt: SetOptions): TSyncPromise<T> {
		const newDefer = this.modify(opt);
		const syncDefer =
			newDefer === undefined
				? TSyncPromise.reject<T>('unsupported')
				: ((Array.isArray(newDefer) ? TSyncPromise.all(newDefer) : TSyncPromise.resolve(newDefer)) as TSyncPromise<T>);
		this.holder.set(this.keyPersist, syncDefer);
		return syncDefer;
	}
}

function createMemoryStore(opt: IOptionReducer<any>): MiddlewareAPI {
	const { key } = opt;
	const reducer = createReducer(opt);
	let state: any = { [key]: {} };
	const ret: MiddlewareAPI = {
		getState() {
			return state;
		},
		dispatch(action: any): any {
			state = reducer(state, action);
			return action;
		},
	};
	return ret;
}



export interface IFetcherOption {
  name?: string;
  setItem?(action: IActionFetch): Record<string, any> | undefined;
}

export type TFetcherFn<T> = IFetcherFunction<T> | ((...args: any[]) => AnyResult<T>);

export interface ICreateOption {
	createStore: typeof createMemoryStore;
	action: string;
	key: string;
}

export function create(opts?: ICreateOption): typeof typeFetcherFn {
	let counter = 0;
	function getName(option?: string | IFetcherOption): string {
		return option === undefined || option === null
			? '' + counter++
			: typeof option === 'string'
				? option
				: getName(option.name);
  }
  function getIterceptor(option?: string | IFetcherOption) {
    return (option && typeof option !== 'string') ? option.setItem : undefined;
  }
	const action = opts ? opts.action : 'action';
	const key = opts ? opts.key : 'local';

  const createStore = (opts && opts.createStore) ? opts.createStore : createMemoryStore;
  const setItemInterceptor: Partial<Record<string, typeof setItem>> = {};
  function setItem(action: IActionFetch): Record<string, any> | undefined {
    const iterceptor = setItemInterceptor[action.name];
    return iterceptor ? iterceptor(action) : undefined;
  }
  const store = createStore({ action, key, setItem });
	function createFetcherImpl<T, SetOptions = any>(
		fns?: TFetcherFn<T>,
		option?: string | IFetcherOption
	): Fetcher<SetOptions, T> {
    const interceptor = getIterceptor(option);
    const name = getName(option);
    if (interceptor) {
      setItemInterceptor[name] = interceptor;
    }
		return new Fetcher<SetOptions, T>(
			{ store, action, key, name },
			!fns ? {} : typeof fns === 'function' ? { load: fns } : fns
		);
	}
	return createFetcherImpl as any;
}
