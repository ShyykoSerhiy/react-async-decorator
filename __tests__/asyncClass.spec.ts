import * as React from 'react';
import { asyncClass } from '../index';
import { create } from '../src/fetcher';

import { renderToString } from 'react-dom/server';
import { createApi } from './helpers';
import { IFetcher0 } from '../src/interfaces';

const createFetcher = create();
@asyncClass
class Test extends React.Component<{ fetcher: IFetcher0<never, string> }, {}> {
	render() {
		const data = this.props.fetcher.get();
		return React.createElement('div', {}, data);
	}
	renderLoader() {
		return React.createElement('div', {}, 'loading');
	}
	renderError(err: Error) {
		return React.createElement('div', {}, err.message);
	}
}

it('render loading', () => {
	const api = createApi<string>();
	const fetcher = createFetcher<never, string>(api.fetch);
	const component = renderToString(React.createElement(Test, { fetcher }));
	expect(component).toMatchSnapshot();
});

it('render error', async () => {
	const api = createApi<string>();
	const fetcher = createFetcher<never, string>(api.fetch);
	api.reject(new Error('custom error'));
	try {
		fetcher.get();
		expect(false).toBeTruthy();
	} catch (e) {
		expect(true).toBeTruthy();
	}
	const catchData = await fetcher.asyncGet().catch(_ => 1);
	expect(1).toBe(catchData);
	const component = renderToString(React.createElement(Test, { fetcher }));
	expect(component).toMatchSnapshot();
});

it('render data', async () => {
	const api = createApi<string>();
	const fetcher = createFetcher<never, string>(api.fetch);
	api.resolve('content');
	try {
		fetcher.get();
		expect(false).toBeTruthy();
	} catch (e) {
		expect(true).toBeTruthy();
	}
	await fetcher.asyncGet();
	const component = renderToString(React.createElement(Test, { fetcher }));
	expect(component).toMatchSnapshot();
});
