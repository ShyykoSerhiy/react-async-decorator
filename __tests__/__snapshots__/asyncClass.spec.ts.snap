// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render data 1`] = `"<div data-reactroot=\\"\\">content</div>"`;

exports[`render error 1`] = `"<div data-reactroot=\\"\\">custom error</div>"`;

exports[`render loading 1`] = `"<div data-reactroot=\\"\\">loading</div>"`;
